### PR TITLE
[[ PlayerPanning ]] Add stereo balance & panning properties to mac player

### PR DESCRIPTION
**Pull request by livecodeian**

### CODE REVIEW

File: docs/dictionary/property/audioPan.lcdoc [hunk 1]
```diff
@@ -0,0 +1,42 @@
+Name: audioPan
+
+Type: property
+
+Syntax: set the audioPan of <player> to <positive/negative percentage>
+
+Summary:
+Specifies the proportion of audio panned from one stereo channel to the other.
+
+Introduced: 9.0.1
```
Comment (montegoulding) [inline docs/dictionary/property/audioPan.lcdoc:10]: @livecodeian Can you add `Associations: player` to the doc

File: docs/dictionary/property/leftBalance.lcdoc [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+Type: property
+
+Syntax: set the leftBalance of <player> to <percentage>
+
+Summary:
+Specifies the volume of the left stereo channel output of a player control.
+
+Introduced: 9.0.1
+
```
Comment (montegoulding) [inline docs/dictionary/property/leftBalance.lcdoc:11]: Again `Associations: player`

File: docs/dictionary/property/audioPan.lcdoc [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+Example:
+set the audioPan of player "myVideoPlayer" to 100 -- 100% of the left channel audio moved to the right.
+
+Value:
+The audioPan is a number between -100 and 100.
+By default, the audioPan property is 0.
+
+Description:
+Use the audioPan property to pan audio from the left & right channels to the other.
```
Comment (montegoulding) [inline docs/dictionary/property/audioPan.lcdoc]: @livecodeian I'm wondering about:
`Use the audioPan property to pan audio from either the left or right channel to the other.`

File: docs/dictionary/property/audioPan.lcdoc [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+Negative values move audio from the right channel to the left.
+Positive values move audio from the left channel to the right.
+Setting the audioPan back to 0 will disable panning.
+
+Important:
```
Comment (montegoulding) [inline docs/dictionary/property/audioPan.lcdoc]: @livecodeian this should be block quote with bold Important:

```
>**Important:** This property only has an effect on media files that contain stereo audio tracks.
> Mono audio tracks cannot currently be panned.
```

File: docs/dictionary/property/leftBalance.lcdoc [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+Value:
+The leftBalance is a number between zero and 100.
+By default, the leftBalance property is 100.
+
+Description:
+Use the leftBalance property to set the volume of audio played through the left speaker.
+
+Setting the leftBalance to zero turns off audio to the left speaker.
+
+Important:
```
Comment (montegoulding) [inline docs/dictionary/property/leftBalance.lcdoc]: @livecodeian see note about this being a block quote above

File: docs/dictionary/property/rightBalance.lcdoc [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+Value:
+The rightBalance is a number between zero and 100.
+By default, the rightBalance property is 100.
+
+Description:
+Use the rightBalance property to set the volume of audio played through the right speaker.
+
+Setting the rightBalance to zero turns off audio to the right speaker.
+
+Important:
```
Comment (montegoulding) [inline docs/dictionary/property/rightBalance.lcdoc]: @livecodeian ditto association and block quote

File: docs/notes/feature-player-pan-mac.md [hunk 7]
```diff
@@ -0,0 +1,8 @@
+# New stereo panning properties added to MacOSX Player object
```
Comment (montegoulding) [inline docs/notes/feature-player-pan-mac.md:1]: Probably not worth worrying about but we should use `macOS`

**Review by montegoulding — CHANGES_REQUESTED** [review]
Hi @livecodeian just a few tweaks to the docs

**Comment by montegoulding** [comment]
@livecode-vulcan review ok dca90c9

**Comment by livecode-vulcan** [comment]
💙 review by @montegoulding ok dca90c93065730c559233f459b13baabc0dceaca

**Comment by montegoulding** [comment]
@livecode-vulcan review not ok dca90c9

@livecodepanos found a crash

**Comment by livecode-vulcan** [comment]
💔 review by @montegoulding not ok dca90c93065730c559233f459b13baabc0dceaca

File: engine/src/mac-av-player.mm [hunk 8]
```diff
@@ -862,6 +896,176 @@ - (void)setPlayer:(AVPlayer *)player
...
+	t_callbacks.clientInfo = p_panning_filter;
+	t_callbacks.init = _audioprocessingtap_init;
+	t_callbacks.prepare = _audioprocessingtap_prepare;
+	t_callbacks.process = _audioprocessingtap_process;
+	t_callbacks.unprepare = _audioprocessingtap_unprepare;
+	t_callbacks.finalize = _audioprocessingtap_finalize;
+
+	MTAudioProcessingTapRef t_tap;
+	OSStatus t_error;
+	t_error = MTAudioProcessingTapCreate(kCFAllocatorDefault, &t_callbacks, kMTAudioProcessingTapCreationFlag_PostEffects, &t_tap);
```
Comment (montegoulding) [inline engine/src/mac-av-player.mm]: @livecodeian we should check t_error here and return false on an error then check it below. At the moment we could be setting r_tap to the uninitialised t_tap here and calling CFRelease on it

**Review by montegoulding — CHANGES_REQUESTED** [review]
@livecodeian sorry I missed that crash. I've just found one more little tweak which looks like a potential crash to me.

**Comment by montegoulding** [comment]
@livecode-vulcan review ok 677392a

**Review by montegoulding — APPROVED** [review]

**Comment by livecode-vulcan** [comment]
💙 review by @montegoulding ok 677392a6da92d0fd4938f4d924c02c6bde21b8c9

**Comment by livecode-vulcan** [comment]
😎 test success 677392a6da92d0fd4938f4d924c02c6bde21b8c9

* try-community-armv6-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-sdk26_ndk16r15/builds/30)
* try-community-armv7-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-armv7-android-sdk26_ndk16r15/builds/22)
* try-community-arm64-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-arm64-android-sdk26_ndk16r15/builds/22)
* try-community-x86-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-x86-android-sdk26_ndk16r15/builds/22)
* try-community-x86_64-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-x86_64-android-sdk26_ndk16r15/builds/22)
* try-community-js-emscripten-sdk1.35: [success](https://vulcan.livecode.com/builders/try-community-js-emscripten-sdk1.35/builds/1544)
* try-community-universal-ios-iphoneos11.4: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos11.4/builds/71)
* try-community-universal-ios-iphonesimulator11.4: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator11.4/builds/71)
* try-community-universal-mac-macosx10.9: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.9/builds/1574)
* try-community-x86-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian8/builds/1568)
* try-community-x86_64-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian8/builds/1564)
* try-community-x86-win32: [success](https://vulcan.livecode.com/builders/try-community-x86-win32/builds/1601)
* try-community-x86_64-win32: [success](https://vulcan.livecode.com/builders/try-community-x86_64-win32/builds/1541)

**Pull request merged**